server/src/test/java/org/elasticsearch/tasks/CancellableTasksTrackerTests.java

```diff
@@ -0,0 +1,174 @@
+                    }
+
+                    if (stateBefore == 2 && stateAfter == 2) {
+                        assertSame(item, getResult);
+                        if (task.getParentTaskId().isSet()) {
+                            assertThat(getByParentResult, hasItem(item));
+                        } else {
+                            assertThat(getByParentResult, empty());
+                        }
+                        assertThat(values, hasItem(item));
+                    }
+
+                    if (stateBefore == 4) {
+                        assertNull(getResult);
+                        if (concurrentRemove == false) {
+                            assertThat(getByParentResult, not(hasItem(item)));
+                        } // else our remove might have completed but the concurrent one hasn't updated the parent ID map yet
+                        assertThat(values, not(hasItem(item)));
+                    }
+                }
+            }, "watch-thread-" + item);
+        }
+
+        void start() {
+            watchThread.start();
+            concurrentRemoveThread.start();
+            actionThread.start();
+        }
+
+        void join() throws InterruptedException {
+            actionThread.join();
+            concurrentRemoveThread.join();
+            watchThread.join();
+        }
+    }
+
+    public void testCancellableTasksTracker() throws InterruptedException {
+
+        final TaskId[] parentTaskIds
+            = randomArray(10, 10, TaskId[]::new, () -> new TaskId(randomAlphaOfLength(5), randomNonNegativeLong()));
+
+        final CancellableTasksTracker<String> tracker = new CancellableTasksTracker<>(new String[0]);
+        final TestTask[] tasks = new TestTask[between(1, 100)];
+
+        final Runnable awaitStart = new Runnable() {
+            private final CyclicBarrier startBarrier = new CyclicBarrier(tasks.length * 3);
+
+            @Override
+            public void run() {
+                try {
+                    startBarrier.await(10, TimeUnit.SECONDS);
+                } catch (InterruptedException | BrokenBarrierException | TimeoutException e) {
+                    throw new AssertionError("unexpected", e);
+                }
+            }
+        };
+
+        for (int i = 0; i < tasks.length; i++) {
+            tasks[i] = new TestTask(
+                new Task(
+                    randomNonNegativeLong(),
+                    randomAlphaOfLength(5),
+                    randomAlphaOfLength(5),
+                    randomAlphaOfLength(5),
+                    rarely() ? TaskId.EMPTY_TASK_ID : randomFrom(parentTaskIds),
+                    Collections.emptyMap()),
+                "item-" + i,
+                tracker,
+                awaitStart
+            );
+        }
+
+        for (TestTask task : tasks) {
+            task.start();
+        }
+
+        for (TestTask task : tasks) {
+            task.join();
+        }
+    }
+}
```

Comment:
I think we should also check the more trivial case where no parent id is set.

Reply:
Maybe I misunderstand, I think we already do?

```
                    rarely() ? TaskId.EMPTY_TASK_ID : randomFrom(parentTaskIds),
```